### PR TITLE
feat: merge common GitHub action steps

### PR DESCRIPTION
**Pull request by ChryssaAliferi**
## Description
This PR refactors GitHub Actions workflows by extracting common logic into reusable components to improve maintainability and reduce code duplication.

**Key Changes:**
- Created a reusable composite action (`setup-and-validate`) that consolidates common CI/CD steps
- Extracted workflow logic into dedicated shell scripts for better maintainability
- Refactored existing workflows (`pr-validation.yml`, `branch-validation.yml`) to use the new common components
- Improved modularity and consistency across CI/CD pipelines

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactor/optimization

## Implementation Details
**New Reusable Action:**
- `.github/actions/setup-and-validate/action.yml` - Composite action that handles Java setup, dependency checks, module detection, quality checks, testing, and AAR building

**Extracted Scripts:**
- `scripts/find-affected-modules.sh` - Determines which modules are affected by changes
- `scripts/run-quality-checks.sh` - Runs Detekt, Android Lint, tests, and builds on specified modules
- `scripts/build-aars.sh` - Builds AAR/JAR files for library modules
- `scripts/send-slack-notification.sh` - Handles Slack notifications with detailed module information

**Refactored Workflows:**
- Updated `pr-validation.yml` and `branch-validation.yml` to use the new composite action
- Maintained all existing functionality while reducing code duplication
- Improved consistency in how modules are detected and processed

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

## How to test?
**Testing the refactored workflows:**
1. **PR Validation:** Create a test PR to verify the `pr-validation.yml` workflow runs correctly with the new composite action
2. **Branch Validation:** Push changes to a feature branch to test the `branch-validation.yml` workflow
3. **Module Detection:** Verify that the `find-affected-modules.sh` script correctly identifies affected modules based on file changes
4. **Quality Checks:** Ensure Detekt, Android Lint, and tests run only on affected modules
5. **AAR Building:** Test that AAR/JAR generation works for library modules when enabled

**Verification Steps:**
- All existing CI/CD functionality should work exactly as before
- Workflows should complete successfully with the same outputs
- Module detection should be accurate and handle edge cases
- Scripts should be executable and handle error conditions gracefully

## Breaking Changes
**No breaking changes**

## Maintainers Checklist

- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Additional Context
**Benefits of this refactoring:**
- **Reduced Code Duplication:** Common CI/CD logic is now centralized in reusable components
- **Improved Maintainability:** Changes to CI/CD logic only need to be made in one place
- **Better Modularity:** Each script has a single responsibility and can be tested independently
- **Enhanced Consistency:** All workflows now use the same logic for module detection and processing
- **Easier Debugging:** Scripts can be run locally for testing and troubleshooting

**Files Changed:**
- **New:** `.github/actions/setup-and-validate/action.yml` (103 lines)
- **New:** `scripts/find-affected-modules.sh` (154 lines)
- **New:** `scripts/run-quality-checks.sh` (165 lines)
- **New:** `scripts/build-aars.sh` (167 lines)
- **New:** `scripts/send-slack-notification.sh` (154 lines)
- **Modified:** `.github/workflows/pr-validation.yml` (significant refactoring)
- **Modified:** `.github/workflows/branch-validation.yml` (renamed and refactored)
- **Removed:** `.github/workflows/branch_validation.yml` (replaced by branch-validation.yml)

### CODE REVIEW

File: .github/workflows/pr-validation.yml [hunk 1]
```diff
@@ -36,196 +36,16 @@ jobs:
           git fetch origin ${{ github.base_ref }}
           git merge origin/${{ github.base_ref }}
 
-      - name: Set up JDK 17
-        uses: actions/setup-java@v4
+      - name: Setup and Validate
+        id: validate
+        uses: ./.github/actions/setup-and-validate
```
Comment (1abhishekpandey) [inline .github/workflows/pr-validation.yml:41]: I can see, we run the same step in the `branch-validation`, refer [here](https://github.com/rudderlabs/rudder-sdk-kotlin/blob/8a149c6476d6b4f2f7fe49bf350de193e2d463d8/.github/workflows/branch-validation.yml#L20).

In this action, we pass this flag: `run_aar_build: 'true'` (whereas in the other action we pass this as `false`).

The current step will execute all the checks again.

Can we plan to extract just the required step into a separate workflow file and just execute that?

It'll improve the execution time and will ensure we don't run any extra step more than once.

File: scripts/run-quality-checks.sh
Comment (1abhishekpandey) [inline scripts/run-quality-checks.sh:1]: This command is not working for the core module:
```
./scripts/run-quality-checks.sh "core" lint
```

Error:

> FAILURE: Build failed with an exception.
What went wrong:
Cannot locate tasks that match ‘core:lint’ as task ‘lint’ not found in project ‘:core’.

File: .github/workflows/pr-validation.yml
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply .github/workflows/pr-validation.yml:1]: I think there’s no downside to running the PR validation check on every pull request.

[the remark this answers is not in the document]
Comment (saikumarrs) [reply .github/workflows/pr-validation.yml:1]: The merge may not always go smooth. A better solution would be to enable GitHub setting that disallows merging PRs if the upstream branch moved ahead.

This way the developer would be responsible to tackle the merge (merge commit or rebase; resolve conflicts) and validations will simply run on the branch.

File: scripts/find-affected-modules.sh [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+# 1. Determines the base branch to compare against
+# 2. Finds changed files
+# 3. Maps changed files to affected modules
+# 4. Handles module dependencies (e.g., if :core is affected, :android is also affected)
+# 5. Outputs results in GitHub Actions format
+
+set -e
+
+# Function to determine base branch
+determine_base_branch() {
```
[the remark this answers is not in the document]
Comment (ChryssaAliferi) [reply scripts/find-affected-modules.sh:16]: You are right on this one. I think we need to use a merge-base commit. Let me think about what we can do.

File: .github/actions/setup-and-validate/action.yml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      shell: bash
+      run: |
+        ./scripts/run-quality-checks.sh "${{ steps.find-modules.outputs.affected_modules }}" "lint"
+
+    - name: Run SonarCloud Analysis
+      if: inputs.sonar_token != ''
+      uses: SonarSource/sonarqube-scan-action@2500896589ef8f7247069a56136f8dc177c27ccf
+      env:
+        GITHUB_TOKEN: ${{ github.token }}
+        SONAR_TOKEN: ${{ inputs.sonar_token }}
```
[the remark this answers is not in the document]
Comment (ChryssaAliferi) [reply .github/actions/setup-and-validate/action.yml:75]: Correct, I will move this to a separate workflow!

File: .github/actions/setup-and-validate/action.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    - name: Find affected modules
+      id: find-modules
+      shell: bash
+      run: |
+        # Make script executable and run it
+        chmod +x scripts/find-affected-modules.sh
+        # Set GITHUB_BASE_REF if provided via base_branch input (for PR context)
+        if [ -n "${{ inputs.base_branch }}" ]; then
+          export GITHUB_BASE_REF="${{ inputs.base_branch }}"
+        fi
```
Comment (1abhishekpandey) [inline .github/actions/setup-and-validate/action.yml:45]: Do we need this `GITHUB_BASE_REF`? I can see we are not using it anywhere.

File: .github/workflows/branch-validation.yml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+on:
+  push:
+    branches:
+      - 'develop'
+      - 'feat/*'
+      - 'fix/*'
+      - 'refactor/*'
+      - 'chore/*'
+      - 'release/*'
+      - 'hotfix/*'
```
Comment (1abhishekpandey) [inline .github/workflows/branch-validation.yml:12]: I think we should also add `ci/*`

```suggestion
on:
  push:
    branches:
      - 'develop'
      - 'feat/*'
      - 'fix/*'
      - 'refactor/*'
      - 'chore/*'
      - 'release/*'
      - 'hotfix/*'
      - 'ci/*'
```

File: .github/workflows/pr-validation.yml [hunk 6]
```diff
@@ -2,28 +2,43 @@ name: PR Validation
...
     steps:
-      - name: Check PR title
-        uses: amannn/action-semantic-pull-request@v5
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      - name: Checkout code
+        uses: actions/checkout@v4
+
+      - name: Check PR Title
+        uses: ./.github/actions/pr-title-check
```
Comment (1abhishekpandey) [inline .github/workflows/pr-validation.yml:18]: I think we can configure the pr-title-check action to run on its own, instead of calling it via different actions.

File: .github/workflows/pr-validation.yml [hunk 7]
```diff
@@ -2,28 +2,43 @@ name: PR Validation
...
+
+  # SonarCloud analysis - runs in parallel, independent of other checks
+  sonarcloud-analysis:
```
Comment (1abhishekpandey) [inline .github/workflows/pr-validation.yml:21]: I think we can configure the SonarCloud action to run on its own, instead of calling it via different actions.

Comment (1abhishekpandey) [reply .github/workflows/pr-validation.yml:21]: I propose the following revamp:
- Refactor the actions to execute directly.
- Ensure each action runs independently.
- Calling scripts is fine, but avoid invoking one action from another to simplify the design and clarify the execution flow.
- We can allow exceptions when necessary, but generally we should avoid this flow to maintain a simplified design.

Let me know your thoughts on this.

File: .github/actions/setup-and-validate/action.yml (hunk 4 above)
Comment (ChryssaAliferi) [reply .github/actions/setup-and-validate/action.yml:45]: I can see that `GITHUB_BASE_REF` is being used in the Slack notification script. The `GITHUB_BASE_REF` is used to show the branch flow in the Slack notification (e.g., "feature-branch → develop"). This provides valuable context about which branch the PR is targeting.

File: .github/workflows/pr-validation.yml (hunk 6 above)
Comment (ChryssaAliferi) [reply .github/workflows/pr-validation.yml:22]: You're right that we could run `pr-title-check` as a separate workflow, but I prefer to keep it within the same workflow file for better organization and to maintain all the validation logic in one place.

By using composite actions, we can still achieve parallelism - the pr-title-check will run as a separate job in parallel with other validation steps, but everything stays organized within the main workflow.

File: .github/workflows/pr-validation.yml (hunk 7 above)
Comment (ChryssaAliferi) [reply .github/workflows/pr-validation.yml:25]: Same as above the above comment ☝️ 

Each action (`pr-title-check`, `sonarcloud-analysis`, etc.) runs completely independently as separate jobs; they are simply organised within the same workflow file for better readability and understanding of the workflow logic.

**Review by 1abhishekpandey — APPROVED** [review]
